### ISSUE TITLE
BGRABitmap and OpenGL.

### ISSUE THREAD

**Issue by fredvs** [bgrabitmap/bgrabitmap#128]
Hello.

I just see that BGRABitmap can deal with OpenGL.
https://wiki.freepascal.org/BGRABitmap_and_OpenGL

Wow.

Do you think it would be possible to draw on a MSEOpenGL component?

In new release of StrumPract: https://github.com/fredvs/strumpract/releases/tag/2.2.0
 (sorry to annoy you always with him but It is used also as a demo-test for MSEgui), if you try "Image Dancer" you will see 3 animations done with BGRABitmap + "OnPaint" and 1 animation done with OpenGL.

All 4 animations are working well but using OpenGL gives a faster and independent animation.

But maybe it is possible to link both: a BGRABitmap design rendered with MSEOpenGL component?

Or this question has no sense because BGRABitmap and OpenGL are too different and in this case, excuse me for the noise.

Fre;D

**Comment by circular17** [comment]
That could be possible. It depends if the same OpenGL units are used. BGRAOpenGL uses GL, GLext, GLU which are part of OpenGL package of FreePascal.

**Comment by fredvs** [comment]
OK, thanks, I will check this asap (but I would not be surprised if Martin did his own tweaked GL, GLext, GLU and then a new bridge would be needed).

Write you later.

Fre;D

**Comment by fredvs** [comment]
> Note that you could anyway transfer the BGRABitmap data to OpenGL as the pixel format is similar.

Ha, so "Fractal Tree" or "Super Formula" code could be used in a OpenGL rendering, without to have to change all the code to be OpenGL compatible?

If yes, that would be marvelous.

Fre;D

**Comment by circular17** [comment]
Yes though that would not be much faster as far as I know. In this case, you have a transfer time from RAM to VRAM and then you get a texture. Then you can draw it many times very fast, or put it on a cube, but that may not be what you're looking for.

**Comment by fredvs** [comment]
OK, I have to study the OpenGL concept.

In fact I would like to have the same result as "Fractal Tree" or "Super Formula" done with BGRABitmap like it is now (and it is great) but with the speed of rendering of the "Dancing Square" OpenGL animation.

By the way, I did lot of search about parallel drawing of canvas and many people said like you that painting canvas is not thread-safe.

**Comment by fredvs** [comment]
> I cannot test your program as I am using MacOS these days.

Hum, I was thinking that MacOS was a Open-Mind OS and was able to run Windows or Linux apps...

;-)

**Comment by fredvs** [comment]
> It depends if the same OpenGL units are used. BGRAOpenGL uses GL, GLext, GLU which are part of OpenGL package of FreePascal.

Aaaargh:

```
unit mseopengl;
...
uses
msegl,mseglext,mseglu, ...
```

**Comment by fredvs** [comment]
I have a surprise for you: ... me!
I do many test with BGRABitmap canvas and I noted that the thing that takes the most time is the creation of the bitmap at each loop:

`TheBitmap := tbgrabitmap.Create(Sender.bounds_cx, Sender.bounds_cy);`
  
So I did try using `TheBitmap` as global variable and only re-paint it at each loop and free it at `form.destroy`.
Doing this makes the animation much faster and independent of other "Onpaint" event.
In fact, very similar to what OpenGL does.

But, would it be possible to assign `TheBitmap.width` and `TheBitmap.height` once the bitmap created?

Because using the same bitmap is fast but if the size of the canvas change, all is not rendered.

Fre;D

**Comment by fredvs** [comment]
Re-re-re Hello.

I think it is the last one!

Yep, I get all what I want, speed and independence for BGRABitmap animation.
Indeed, a global variable do the trick and for the resizing of the canvas:
So, the bitmap created at form.oncreated and in the anim-loop:

```
while true do
begin
if (TheBitmap.width <> Sender.bounds_cx) or (TheBitmap.width <> Sender.bounds_cx)
then
begin 
 Thebitmap.free;
 TheBitmap := tbgrabitmap.Create(Sender.bounds_cx, Sender.bounds_cy);
 end;
...
```

Perfect, perfect and maybe the best news for you: I will really close my comments.

Thank to all.

Fre;D

**Comment by fredvs** [comment]
Now with the full speed of "Fractal Tree" and "Super Fomula" even in full-screen, there is a other problem.

It is completely hypnotizing and difficult to not feel in trance.

But ok, I stop.

Wow for BGRABitmap.

**Comment by circular17** [comment]
You can do:
`TheBitmap.SetSize(Sender.bounds_cx, Sender.bounds_cy);`

If the size doesn't change, it won't do anything.

**Comment by fredvs** [comment]
> TheBitmap.SetSize(Sender.bounds_cx, Sender.bounds_cy);

Ooops, just when I released ver 2.2.2 because the previous version was not fear.
OK, it is even more easy.

Many thanks and big wow.

**Comment by fredvs** [comment]
Hello.
Using this works like charm:

```
if (Bitmap.width <> Sender.bounds_cx) or (Bitmap.height <> Sender.bounds_cy)
then
 bitmap.SetSize(Sender.bounds_cx, Sender.bounds_cy);
```

Ok, I will fast update the assets of ver 2.2.2.
